lib/internal/options.js

```diff
@@ -47,15 +48,16 @@ function getOptionValue(optionName) {
 }
 
 function getAllowUnauthorized() {
-  const allowUnauthorized = process.env.NODE_TLS_REJECT_UNAUTHORIZED === '0';
-
-  if (allowUnauthorized && warnOnAllowUnauthorized) {
+  if (warnOnAllowUnauthorized) {
     warnOnAllowUnauthorized = false;
-    process.emitWarning(
-      'Setting the NODE_TLS_REJECT_UNAUTHORIZED ' +
-      'environment variable to \'0\' makes TLS connections ' +
-      'and HTTPS requests insecure by disabling ' +
-      'certificate verification.');
+    allowUnauthorized = process.env.NODE_TLS_REJECT_UNAUTHORIZED === '0';
+    if (allowUnauthorized) {
```

Comment:
`warnOnAllowUnauthorized` should only be set `false` when both conditions are `true`, otherwise we will not emit the warning when the `NODE_TLS_REJECT_UNAUTHORIZED` is turned on.

Reply:
But this is exactly what it does?!

on initial warnOnAllowUnauthorized is set to true. so the if in line 51 is entered. we set warnOnAllowUnauthorized to false, as we are only interested into warning once. We check if allowUnauthorized is set to 0, and cache it.. If it is true, we enter the if block in line 54. we emit the warning. anyhow we cached the value and return it.

The second time we cal getAllowUnauthorized warnOnAllowUnauthorized is set to false. We dont need to check anything else as we dont need to emit a warning anymore. we return the cached result.

ANd imho NODE_TLS_REJECT_UNAUTHORIZED should only be set, at start up and not at runtime, as this could mean, that a malicious package could turn off the tls check.

Reply:
> we set warnOnAllowUnauthorized to false, as we are only interested into warning once.

This code basically emits a warning when that flag is enabled for the first time, whether it is during startup or runtime.

> ANd imho NODE_TLS_REJECT_UNAUTHORIZED should only be set, at start up and not at runtime

Ideally, yes, practically, people do things very differently than we expect, anyway, I don't feel comfortable approving in the current way (breaking the old behavior) but let's see if we can get more people who have more knowledge on this topic.

Somehow, this is also something related to security, so I don't think hiding the warning when the flag is enabled runtime is a good thing.